providers/loopia/loopiaProvider.go

```diff
@@ -0,0 +1,426 @@
+	providers.DocOfficiallySupported: providers.Cannot(),
+}
+
+// Section 2: Define the API client.
+
+// See client.go
+
+// newDsp generates a DNS Service Provider client handle.
+func newDsp(conf map[string]string, metadata json.RawMessage) (providers.DNSServiceProvider, error) {
+	return newHelper(conf, metadata)
+}
+
+// newReg generates a Registrar Provider client handle.
+func newReg(conf map[string]string) (providers.Registrar, error) {
+	return newHelper(conf, nil)
+}
+
+// newHelper generates a handle.
+func newHelper(m map[string]string, metadata json.RawMessage) (*LoopiaClient, error) {
+	if m["username"] == "" {
+		return nil, fmt.Errorf("missing Loopia API username")
+	}
+	if m["password"] == "" {
+		return nil, fmt.Errorf("missing Loopia API password")
+	}
+
+	dbg := false
+	var err error
+	if m["debug"] != "" { //debug is optional
+		dbg, err = strconv.ParseBool(m["debug"])
+		if err != nil {
+			return nil, fmt.Errorf("creds.json requires the debug setting as a 'string': 't', 'true', 'True' etc.")
+		}
+	}
+	api := NewClient(m["username"], m["password"], dbg)
+
+	quota := m["rate_limit_per"]
+	err = api.requestRateLimiter.setRateLimitPer(quota)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected value for rate_limit_per: %w", err)
+	}
+	return api, nil
+}
+
+// Section 3: Domain Service Provider (DSP) related functions
+
+// // ListZones lists the zones on this account.
+func (c *LoopiaClient) ListZones() ([]string, error) {
+
+	listResp, err := c.GetDomains()
+	if err != nil {
+		return nil, err
+	}
+
+	zones := make([]string, len(listResp))
+	if c.Debug {
+		fmt.Printf("DEBUG: DOMAIN LIST START\n")
+	}
+	for i, zone := range listResp {
+		for _, prop := range zone.Properties {
+			if prop.Name() == "domain" { // the zones name is stored in property 'domain'
+				if c.Debug {
+					fmt.Printf("DEBUG: DOMAIN LIST %d: %v\n", i, prop.String())
+				}
+				// zone := zone
+				zones[i] = prop.String()
+			}
+		}
+	}
+	if c.Debug {
+		fmt.Printf("DEBUG: DOMAIN LIST END\n")
+	}
+	return zones, nil
+}
+
+// NB(tal): To future-proof your code, all new providers should
+// implement GetDomainCorrections exactly as you see here
+// (byte-for-byte the same). In 3.0
+// we plan on using just the individual calls to GetZoneRecords,
+// PostProcessRecords, and so on.
+//
+// Currently every provider does things differently, which prevents
+// us from doing things like using GetZoneRecords() of a provider
+// to make convertzone work with all providers.
+
+// GetDomainCorrections get the current and existing records,
+// post-process them, and generate corrections.
+func (client *LoopiaClient) GetDomainCorrections(dc *models.DomainConfig) ([]*models.Correction, error) {
+	existing, err := client.GetZoneRecords(dc.Name)
+	if err != nil {
+		return nil, err
+	}
+	models.PostProcessRecords(existing)
+	clean := PrepFoundRecords(existing)
+	PrepDesiredRecords(dc)
+	return client.GenerateDomainCorrections(dc, clean)
```

Comment:
I haven't documented this but... please rename "GenerateDomainCorrections" to "GenerateZoneRecordsCorrections".  Thanks!

Reply:
OK.